credentials/tls/certprovider/store.go

```diff
@@ -79,42 +83,107 @@ func GetProvider(name string, config interface{}) (Provider, error) {
 	if err != nil {
 		return nil, err
 	}
-
-	sk := storeKey{
+	pk := providerKey{
 		name:   name,
 		config: string(stableConfig.Canonical()),
 	}
-	if wp, ok := provStore.providers[sk]; ok {
-		wp.refCount++
-		return wp, nil
+
+	// Create a new provider for the (name+config) combo or use an existing one.
+	var wp *wrappedProvider
+	wp, ok := provStore.providers[pk]
+	if !ok {
+		provider := builder.Build(stableConfig)
+		if provider == nil {
+			return nil, fmt.Errorf("certprovider.Build(%v) failed", pk)
+		}
+		wp = &wrappedProvider{
+			prov:     provider,
+			storeKey: pk,
+			store:    provStore,
+		}
+		provStore.providers[pk] = wp
 	}
 
-	provider := builder.Build(stableConfig)
-	if provider == nil {
-		return nil, fmt.Errorf("certprovider.Build(%v) failed", sk)
+	// Create a new distributor for the (name+config+opts) combo or use an
+	// existing one. The reference count for the provider is incremented only
+	// when creating a new distributor.
+	dk := distributorKey{
+		name:    name,
+		config:  string(stableConfig.Canonical()),
+		options: opts,
 	}
-	wp := &wrappedProvider{
-		Provider: provider,
-		refCount: 1,
-		storeKey: sk,
-		store:    provStore,
+	if wd, ok := provStore.distributors[dk]; ok {
```

Comment:
Can you do this lookup first, and save the provider lookup if a distributor already exists?

Reply:
Done. Thanks.